gateway/gateway.go

```diff
@@ -787,6 +788,75 @@ func (gateway *HandleT) pendingEventsHandler(w http.ResponseWriter, r *http.Requ
 	w.Write([]byte(fmt.Sprintf("{ \"pending_events\": %d }", totalPendingCount)))
 }
 
+type failedEventsRequestPayload struct {
+	TaskRunID string `json:"task_run_id"`
+}
+
+func (gateway *HandleT) fetchFailedEventsHandler(w http.ResponseWriter, r *http.Request) {
+	gateway.failedEventsHandler(w, r, "fetch")
+}
+
+func (gateway *HandleT) clearFailedEventsHandler(w http.ResponseWriter, r *http.Request) {
+	gateway.failedEventsHandler(w, r, "clear")
+}
+
+func (gateway *HandleT) failedEventsHandler(w http.ResponseWriter, r *http.Request, reqType string) {
+	gateway.logger.LogRequest(r)
+	atomic.AddUint64(&gateway.recvCount, 1)
+	var errorMessage string
+	defer func() {
+		if errorMessage != "" {
+			gateway.logger.Debug(errorMessage)
+			http.Error(w, response.GetStatus(errorMessage), 400)
+		}
+	}()
+
+	payload, _, err := gateway.getPayloadAndWriteKey(w, r, reqType)
+	if err != nil {
+		errorMessage = err.Error()
+		return
+	}
+
+	if !gjson.ValidBytes(payload) {
+		errorMessage = response.GetStatus(response.InvalidJSON)
+		return
+	}
+
+	var reqPayload failedEventsRequestPayload
+	err = json.Unmarshal(payload, &reqPayload)
+	if err != nil {
+		errorMessage = err.Error()
+		return
+	}
+
```

Comment:
Can we add all the lines above till here into a helper function since the same code is being used in clearhandler too